core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -1015,6 +1015,11 @@ public Builder upgradeFormatVersion(int newFormatVersion) {
           "Cannot downgrade v%s table to v%s",
           formatVersion,
           newFormatVersion);
+      Preconditions.checkArgument(
+          newFormatVersion <= formatVersion + 1,
+          "Cannot skip format version(s) to upgrade v%s table to v%s",
+          formatVersion,
+          newFormatVersion);
```

Comment:
Curious, what's the rationale behind preventing upgrades from 1 to 3 (or more generally skipping versions)? 

I think it should technically possible to enable that upgrade safely (let me know if I'm missing something there). Also, with this check if  someone who really wants to "skip", they would just upgrade to 2 and then right to 3 after that since there's no real additional logic. Subsequent writes would write out the new format's metadata after that.

For what it's worth, I don't think it'll be common so maybe this check doesn't really matter in practice.

Reply:
I guess we could allow this and just internally do the upgrade to 2 first? My only goal here would be to make sure that we don't support 2 different upgrade paths. 

` 1 -> 3` and `1 -> 2 -> 3`

But if upgrading 1 ->3 is just syntatic sugar for 1 -> 2 -> 3 I don't see the issue

Reply:
Ok looking through the rest of this update I think it probably shouldn't be disallowed to skip versions, we should just be internally doing the stepwise upgrade.

Because of that I think we need to remove all the gates on skipping versions but also all of the added tests should not assume that the "version skip" is just a single version, and instead just test skipping to latest. 

So basically my thoughts for testing strategy is
Test
```
N => N+ 1
And
N => Max_Supported Version
```
Thoughts?

Reply:
> My only goal here would be to make sure that we don't support 2 different upgrade paths.

Thanks for this clarification, that makes sense to me. Having a single upgrade path makes it easier to rationalize about any issues that come up for an update (we don't have to think about "Oh does X happen when it's 1 to 3 directly or something else". It's a single path and so in case of any issues it must be at some specific point in the path). 

I think I'm overall good with that, as long as we can keep the combinatorics to a minimum for testing. Which I think your strategy is as good as it gets?

Reply:
> My only goal here would be to make sure that we don't support 2 different upgrade paths.

Thanks for this clarification, that makes sense to me. Having a single upgrade path makes it easier to rationalize about any issues that come up for an update (we don't have to think about "Oh does X happen when it's 1 to 3 directly or something else". It's a single path and so in case of any issues it must be at some specific point in the path). 

I think I'm overall good with that, as long as we can keep the combinatorics to a minimum for testing. Which I think your strategy is as good as it gets?

Reply:
This sounds alright, I'll remove the assert here and find a way to perform a serial upgrade internally.

Some implementation notes I have so far:
- Currently the internal implementation of performing a format version upgrade is to add to the list of metadata changes a MetadataUpdate.UpgradeFormatVersion instance.
- `MetadataUpdate.UpgradeFormatVersion` tracks the target version but not the origin version. This may mean that multiple instances of MetadataUpdate.UpgradeFormatVersion should be populated into the list of metadata changes to simulate this multi-version upgrade.
- `UpdateRequirements` class which defines all upgrade requirements has no requirements for the `MetadataUpdate.UpgradeFormatVersion` class due to the fact that backwards compatibility has never been broken (yet).

Reply:
Also added additional asserts in TestFormatVersions to check that appropriate `MetadataUpdate.UpgradeFormatVersion` instances are registered in changes correctly.